src/software/amazon/ionschema/internal/TypeReference.kt

```diff
@@ -0,0 +1,98 @@
+package software.amazon.ionschema.internal
+
+import software.amazon.ion.IonStruct
+import software.amazon.ion.IonSymbol
+import software.amazon.ion.IonText
+import software.amazon.ion.IonValue
+import software.amazon.ionschema.InvalidSchemaException
+import software.amazon.ionschema.Schema
+
+/**
+ * Provides a factory method that translates an ISL type reference into a function
+ * that returns a Type instance.
+ *
+ * Types that can't be resolved yet are instantiated as [TypeReferenceDeferred] objects
+ * that are resolved by [SchemaImpl.resolveDeferredTypeReferences] prior to asserting
+ * that the schema is valid.
+ */
+internal class TypeReference private constructor() {
+    companion object {
+        fun create(ion: IonValue, schema: Schema, isField: Boolean = false): () -> TypeInternal {
+            var type: TypeInternal? = null
+            if (!ion.isNullValue) {
+                type = when (ion) {
+                    is IonStruct -> {
+                        val id = ion.get("id") as? IonText
+                        if (id != null) {
```

Comment:
favor `when` instead of long `if/else` chains

```kotlin 
when {
  id != null -> {}
  isField -> {}
  ion.size() ==1 && ion.get("type") != null -> {}
  else -> {}
}
```

Reply:
hey, didn't realize that was allowed, nice!  updated